### ISSUE TITLE
Matchers can't compare promises

### ISSUE THREAD

**Issue by chrisgodsey** [jasmine/jasmine#1543]
So I've got some long standing unit tests that have been running for several months that are all just verifying that the return of functions is wrapped in a promise, and are all basically iterations of:

expect(object.function()).toEqual(Promise.resolve());

These tests run fine on jasmine 2.8, but if I update to 2.99 (or 3.0+), the tests fail with the message:

Error: Expected [object Promise] to equal [object Promise].

I couldn't see any recent checks in the toEqual function, so I'm wondering if I'm missing something?

**Comment by sgravrock** [comment]
Chances are that your tests would have always passed prior to 2.9.0, even if the returned promise was rejected or resolved to a different value than expected. 

In general, `.toEqual` tries to determine if two objects represent the same value. That's not really possible with promises because they don't expose any properties (at least not in a way that's either public or portable). Prior to 2.9.0, `.toEqual` treated any two promises as equal because they had the same constructor and had the same properties. Starting with 2.9.0, `.toEqual` treats two promises as equal only if they are the same object, i.e. `===` equality.

It looks like you're trying to assert that the promise returned by `object.function()` has been resolved to `undefined`. There isn't a way to do exactly that, but you can assert that it *eventually* resolves to `undefined`:

```
it('should return a promise that resolves to undefined', function() {
   return object.function().then(function (result) {
      expect(result).toBeUndefined();
   });
});
```

**Comment by chrisgodsey** [comment]
Yeah, I had been changing everything to this as an attempt to fix things:

expect(object.function().then).toBeTruthy();

Thanks for the response, I'll just update the tests.

**Comment by chrisgodsey** [comment]
So following up on this - why does toEqual treat promises as equal only if they're the same object? Isn't this the function of toBe?

I can expect({}).toBe({}) and I will get a failure, but if I do expect({}).toEqual({}) I will get a pass.

**Comment by sgravrock** [comment]
> why does toEqual treat promises as equal only if they're the same object? 

Because it's not possible (as far as I know) for `toEqual` to do the same thing with promises that it does with other objects. Let's set aside the pending and rejected cases for now and say that two promises should be considered equal if they're both resolved to the same value. The comparison logic would look something like this:

```
// ...
if (isPromise(a) && isPromise(b) {
   return isResolved(a) && isResolved(b) && equals(resolvedValue(a), resolvedValue(b));
}
// ...
```

The problem is that [it's not possible to implement `isResolved` or `resolvedValue`](https://stackoverflow.com/questions/30564053/how-can-i-synchronously-determine-a-javascript-promises-state). The Promise spec doesn't provide any synchronous API for inspecting the state of the promise, only the async `then` method. Since `toEqual` is necessarily synchronous, it can't use those async methods. If you're not convinced, you can prove this to yourself by trying to implement `isResolved` and `resolvedValue`. Make sure that they return a boolean and use only features that are documented in the Promise spec.

So that leaves only a few options:
1. Make matchers asynchronous. This would be a major effort and would require literally every Jasmine spec out there to be rewritten.
2. Throw an exception if a promise is passed to `toEqual`.
3. Treat all promises as equal, as before 2.9.0.
4. Treat promises as equal if they are the same instance, as in 2.9.0.

Option 1 is out of the question, and I think that option 4 is significantly more useful than 2 or 3.

**Comment by chrisgodsey** [comment]
Yeah, I totally understand why using toEqual for two promises doesn't check if they resolve the same value - I was just more remarking that just making toEqual and toBe be the same thing (but only for promises) is a confusing interaction.

e.g expect(Promise.resolve()).toEqual(Promise.resolve()) = fail, expect({}).toEqual({}) = pass is confusing when expect(Promise.resolve()).toBe(Promise.resolve()) = fail and expect({}).toBe({}) = fail

This unique implementation of toEqual for promises doesn't gain you any functionality you don't already have, it's just conflating what toEqual actually does.

**Comment by slackersoft** [comment]
Jasmine's `toEqual` is a best effort deep compare of two objects. Other places where it also falls back to work the same as `toBe` are for literal values like numbers and strings. The only really viable options for comparing two promises are (per @sgravrock's [note](https://github.com/jasmine/jasmine/issues/1543#issuecomment-382140206)):

1. Don't allow it. Throw an `Error` if we ever find a Promise anywhere in the object graph of two things being compared.
2. All promises are equal as before 2.9.0. This breaks previously fixed issues (#1314) and also means that `expect(Promise.resolve()).toEqual(Promise.reject())` would also pass in your tests.
3. Fall back to the best effort of checking whether they are the same object, as currently.

Of these options, the current (number 3) seems like the best option since it allows some form of equality checking, which is especially useful for spy argument matching, without reporting false equivalency for two Promises that are easily recognizable to a human as being different things.

In general Jasmine believes that a spec that should *fail* but *passes* is _much_ worse than a spec that should *pass* but *fails* as the former is more likely to hide a real issue in the underlying code that the latter.

We would be happy to hear about any other options that you (or any other reader) have to more accurately compare to Promises synchronously without having to know which Promise library might be in use.

Hope this helps. Thanks for using Jasmine!

**Comment by slackersoft** [comment]
Jasmine 3.1 introduced the new [`expectAsync`](https://jasmine.github.io/api/edge/global.html#expectAsync) syntax which should allow better matching of the value that a Promise resolves to. Jasmine 3.3 makes this work with non-built-in versions of Promises as well.

This new syntax should address the underlying issue here, so I'm going to close this.
